Line editing layered_store (#2332)
mirage/irmin: doc/irmin-pack/design/layered_store.md

﻿# Irmin-Pack Layered Store/GC Design Document

Date: 2022-02-03

## Introduction

This is a short document to describe the design of a garbage
collection mechanism for `irmin-pack` (the variant of Irmin currently
used by Tezos).

**TL;DR**

> The `irmin-pack` file stores objects. References from one object to

Reviewer: ```suggestion
A pack file stores Irmin's objects. References from one object to
```
The name "pack file" describes a file which packs multiple objects (so `irmin-pack` got his name from that way of organizing the data on disk). The previous sentence was ambiguous as you noticed, but I'm afraid that "the `irmin-pack` file" reformulation would carry a different meaning than intended... I'm not sure if my suggestion is clearer, let me know if it's still confusing or if you think of a better way to phrase it!

Author: Perfect!

  disk I/O. The OS caches should instead remain allocated to the main
  Tezos node. In particular, the worker should not disturb the caches
  for the main process when the worker calculates the reachable
  objects for the given commit.
- With the exception of the calculation of the reachable objects from
  a commit, the worker is **simply copying bytes between files.** This
  makes the worker very simple and hopefully makes it easy to
  optimise performance and ensure correctness of operation.

### Crash Correctness

We now describe what happens in the event of a system crash. The
interesting cases are when the worker is active or has just
terminated.

If the worker process is active, then it may have created part or all
of the next object-store and next suffix. However, the control file
will still point to the current object store and suffix. On restart,
the main process opens the control file. At this point it knows which
files are the "real" object-store and suffix (those listed in the
control file). The main process can delete any other files it finds (which will be
the partial files the worker was creating).

> If the worker process has just terminated and a crash occurs, then the

Author: Li 299 doesn't have a complete sentence. Could you clarify please?

Reviewer: Hm there's an extra empty line 298 (which needs to be removed) but the sentence continues below. Does the following makes more sense?
> If the worker process terminates just before a crash occurs, then either the current `control` file contains the atomically renamed `control.nnnn` worker's file, or it has been left untouched by the worker. If it was successfully renamed, then the main process will start operating on the new object-store and suffix file created by the worker (and will delete any residual old version of those files).

Author: Thank you! I fixed it and (hopefully) clarified two of those sentences. Please ensure they're still accuarte.

> `control` (which was successfully renamed from `control.nnnn` ), and

Author: Should li. 299 read: `control.nnnn` over the current `control` was recorded persistently on
disk or it wasn't. 

Would you please clarify this sentence? It doesn't make sense as it's written.

Reviewer: (marked as resolved because it's easier to discuss in your previous comment that contains the rest of the sentence)
